### PR TITLE
feat: add version bumping, solution listing, and API path validation

### LINKED ISSUE #321: feat(cli): build --bump for version increment and catalog push --latest
Building and publishing solutions requires manually tracking and specifying version numbers. This is error-prone and tedious.

**Proposed -- two independent features**:

### 1. `build solution --bump <level>`

Automatically increment the version based on the last published version:

```bash
# Current latest: 1.2.3
cldctl build solution --bump patch   # builds 1.2.4
cldctl build solution --bump minor   # builds 1.3.0
cldctl build solution --bump major   # builds 2.0.0
```

- Reads the current latest version from the catalog
- Increments according to semver rules
- Passes the computed version to the build
- Fails if no previous version exists (use `--version` for first publish)

### 2. `catalog push --latest`

Tag the pushed version as the `latest` alias in the catalog:

```bash
cldctl catalog push my-solution@1.2.4 --latest
```

- Sets the `latest` pointer so consumers get this version by default
- Without `--latest`, the version is pushed but not promoted

These are independent: `--bump` is a build concern, `--latest` is a publish concern. They compose naturally:

```bash
cldctl build solution --bump patch
cldctl catalog push my-solution@1.2.4 --latest
```

**Complexity**: S

### LINKED ISSUE #346: ux: get solution vs get provider inconsistent behavior
# get solution vs get provider: inconsistent behavior

**Type**: enhancement
**Complexity**: S
**Component**: scafctl CLI -- `get` subcommands

## Problem

`scafctl get provider` and `scafctl get solution` are both under the `get` verb but behave differently:

- `scafctl get provider` without arguments lists all registered providers. With a name argument, it shows details for one provider.
- `scafctl get solution` without arguments auto-discovers and displays a single local solution file. With a name argument, it fetches one solution from the catalog.

This asymmetry is confusing because the `get` verb implies a consistent pattern. A user who learns `get providers` lists all providers will expect `get solutions` to list all solutions.

The underlying reason is that providers live in an internal registry (finite, enumerable), while solutions live in files and catalogs (distributed, not enumerable from one place). But that distinction is not obvious from the command surface.

## Expected Behavior

`scafctl get solution` without arguments should either:

1. List solutions from the catalog (similar to `catalog list`) -- consistent with `get provider` listing all providers
2. Keep auto-discovery but make the asymmetry explicit in help text and error messages

Option 1 is more consistent. Option 2 is the minimum fix.

If option 1 is chosen, auto-discovery could move to a `--local` flag or remain as a fallback when no catalog is configured.

## Current Behavior

~~~bash
# Lists all providers
scafctl get provider

# Auto-discovers and displays ONE local solution
scafctl get solution

# These feel like they should be symmetric but are not
~~~

## Scope

S -- CLI command behavior change or improved help text. Likely touches:

- `get solution` command handler
- help text for `get solution`
- possibly catalog integration for listing

### LINKED ISSUE #375: bug(api): /v1/solutions/run rejects catalog names and OCI URLs
## Summary

The `/v1/solutions/run` API endpoint enforces HTTP(S)-only URL validation on the `path` field, rejecting catalog names and OCI URLs with `"path must be an HTTP or HTTPS URL"`. The CLI resolves all of these formats correctly via the engine's resolution logic, but the API server blocks them at the validation layer.

## Reproduction

```
POST /v1/solutions/run {"path": "entra-user-groups"}
  -> 400 "solution-run: path must be an HTTP or HTTPS URL"

POST /v1/solutions/run {"path": "oci://ghcr.io/..."}
  -> 400 "solution-run: path must be an HTTP or HTTPS URL"

POST /v1/solutions/run {"path": "https://...solution.yaml"}
  -> works

CLI: scafctl run resolver entra-user-groups
  -> works (resolves from catalog)
```

## Root Cause

`requireURLPath()` in `pkg/api/endpoints/solutions.go` rejects anything not starting with `http://` or `https://`:

```go
func requireURLPath(path, opName string) error {
    lower := strings.ToLower(path)
    if !strings.HasPrefix(lower, "http://") && !strings.HasPrefix(lower, "https://") {
        return huma.NewError(http.StatusBadRequest,
            fmt.Sprintf("%s: path must be an HTTP or HTTPS URL", opName))
    }
    return nil
}
```

The engine already supports catalog name, OCI URL, and HTTP URL resolution. The API server just needs to stop blocking valid inputs.

## Suggested Fix

Replace the HTTP-only check with validation that accepts the formats the engine supports:

1. **Catalog names** (bare names like `entra-user-groups`) -- safe, resolved via catalog registry
2. **OCI URLs** (`oci://...`) -- safe, resolved via OCI registry client
3. **HTTP/HTTPS URLs** -- already allowed

**Local file paths** should remain blocked in the API server (path traversal risk) unless explicitly opted in via a server flag (e.g., `--allow-local-paths`). The `/v1/solutions/render` endpoint uses the same `requireURLPath()` and likely needs the same fix.

## Affected Areas

- `pkg/api/endpoints/solutions.go` -- `requireURLPath()` function
- Possibly `/v1/solutions/render` endpoint (same validation)

## Scope

**S** -- validation function change + test updates.

### PR DESCRIPTION
**Pull request by abaker-9**
- add --bump flag (patch/minor/major) to build solution for auto-incrementing version based on last published version
- add --latest flag to catalog push for tagging remote artifacts
- change get solution default to list catalog solutions; add --local flag for previous auto-discovery behavior
- replace requireURLPath with requireRemotePath to accept OCI URLs and catalog names while blocking local paths and file://
- add ErrNoVersions sentinel to eliminate double catalog listing in NextPatchVersion
- detect --bump conflict with versioned --tag after tag parsing
- sort solution list output for deterministic ordering
- strip stale git annotations when repo is clean or --no-git-metadata is used

BREAKING CHANGE: get solution without arguments now lists catalog solutions instead of auto-discovering a local file. Use --local for the previous behavior.

Closes #375
Closes #346
Closes #321

### CODE REVIEW

File: pkg/api/endpoints/solutions.go [hunk 1]
```diff
@@ -109,16 +109,44 @@ type SolutionTestResponse struct {
...
+	// Block all other URL schemes (file://, ftp://, data://, etc.).
+	if strings.Contains(lower, "://") {
+		return huma.NewError(http.StatusBadRequest,
+			fmt.Sprintf("%s: unsupported URL scheme; only http://, https://, and oci:// are allowed", opName))
+	}
+
+	// Block obvious local-path indicators.
+	if strings.HasPrefix(path, "/") || strings.HasPrefix(path, ".") ||
+		strings.HasPrefix(path, "~") || strings.Contains(path, "\\") ||
+		scafpath.HasWindowsDrivePrefix(path) {
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/api/endpoints/solutions.go:146]: Fixed. requireRemotePath now blocks paths with `/` where the first segment lacks `.` or `:` (mirrors IsCatalogReference logic). Blocks `configs/secret` and `dir/../../etc/passwd` while allowing `ghcr.io/org/solution`. Added test cases. Thanks!

File: pkg/cmd/scafctl/catalog/push.go [hunk 2]
```diff
@@ -331,6 +336,20 @@ func runPush(ctx context.Context, opts *PushOptions) error {
 		}
 	}
 
+	// Tag as "latest" in the remote registry if requested.
+	// This intentionally bypasses catalog.ValidateAlias (which rejects "latest")
+	// because push --latest is a publish-time concern, not a manual alias operation.
+	if opts.Latest {
+		w.Infof("Tagging %s@%s as \"latest\"...", displayName, ref.Version.String())
+		_, tagErr := remoteCatalog.Tag(ctx, ref, "latest")
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/catalog/push.go]: Fixed. When `--as` is used, the `--latest` tagging now uses a copy of `ref` with `Name` set to the target name, so `Tag()` targets the correct remote repository. Thanks!

File: pkg/api/endpoints/solutions.go [hunk 3]
```diff
@@ -109,16 +109,44 @@ type SolutionTestResponse struct {
...
+	// Bare filenames ending in .yaml/.yml/.json are almost certainly local files.
+	if strings.HasSuffix(lower, ".yaml") || strings.HasSuffix(lower, ".yml") || strings.HasSuffix(lower, ".json") {
 		return huma.NewError(http.StatusBadRequest,
-			fmt.Sprintf("%s: path must be an HTTP or HTTPS URL", opName))
+			fmt.Sprintf("%s: path must be a URL or catalog reference, not a local file path", opName))
 	}
+
+	// Anything else is treated as a catalog reference (e.g. "my-app",
+	// "my-app@1.0.0", "ghcr.io/org/solution").
 	return nil
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/api/endpoints/solutions.go:171]: Fixed in the same change that addressed the earlier P1 thread. `requireRemotePath` now blocks paths containing `/` where the first segment lacks `.` or `:` (i.e., not a registry hostname). This blocks `configs/solution`, `a/../b`, and `dir/../../etc/passwd` while allowing `ghcr.io/org/solution` and `localhost:5000/solution`. Added test cases for all three patterns.

File: pkg/api/endpoints/solutions.go [hunk 4]
```diff
@@ -145,7 +173,7 @@ func RegisterSolutionEndpoints(humaAPI huma.API, hctx *api.HandlerContext, prefi
 		Description: "Validates a solution file and returns findings (errors, warnings, info).",
 		Tags:        []string{"Solutions"},
 	}, hctx, http.StatusOK), func(ctx context.Context, input *SolutionLintRequest) (*SolutionLintResponse, error) {
-		if err := requireURLPath(input.Body.Path, "solution-lint"); err != nil {
+		if err := requireRemotePath(input.Body.Path, "solution-lint"); err != nil {
 			return nil, err
 		}
 		sol, err := inspect.LoadSolution(ctx, input.Body.Path)
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/api/endpoints/solutions.go:200]: The `requireRemotePath` function is thoroughly covered by unit tests in `solutions_test.go` (allowed paths, blocked paths, catalog name via run endpoint, path traversal, and relative directory cases). Since `requireRemotePath` is a pure validation function called identically by all six endpoints, unit tests provide sufficient regression coverage here. Adding integration tests for each endpoint would duplicate the same validation logic. Happy to discuss further.

File: pkg/cmd/scafctl/get/solution/solution.go [hunk 5]
```diff
@@ -79,13 +106,20 @@ func CommandSolution(cliParams *settings.Run, ioStreams *terminal.IOStreams, pat
 
 				return exitcode.WithCode(err, exitcode.InvalidInput)
 			}
+
+			// No args, no --file, and not --local: list solutions from catalog
+			if len(args) == 0 && options.File == "" && !options.Local {
+				return options.ListSolutions(ctx)
+			}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/get/solution/solution.go:113]: Added `TestCommandSolution_NoArgsCallsListSolutions` which executes the command with no args and asserts it reaches `ListSolutions` (producing the "No solutions found" message in a test environment without a catalog). This exercises the `RunE` branch at line 113.

File: pkg/cmd/scafctl/catalog/push.go [hunk 6]
```diff
@@ -331,6 +336,20 @@ func runPush(ctx context.Context, opts *PushOptions) error {
...
+		if tagErr != nil {
+			w.Warningf("failed to tag as latest: %v", tagErr)
+			// Non-fatal: the push itself succeeded
+		} else {
+			w.Successf("Tagged %s@%s as \"latest\"", displayName, ref.Version.String())
+		}
+	}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/catalog/push.go:355]: The `--latest` dry-run output is already present in `runPush` (line 261: `w.Infof("  --latest: would tag as \"latest\"")`) but testing it requires a real local catalog artifact to reach the dry-run branch (lines 237-247 resolve the artifact first). This is integration-test territory. The flag registration is already tested in `TestCommandPush_LatestFlag`. Adding an integration test with a temp catalog is out of scope for this PR but tracked for follow-up.

File: pkg/cmd/scafctl/build/solution.go [hunk 7]
```diff
@@ -297,24 +318,43 @@ func runBuildSolution(ctx context.Context, opts *SolutionOptions) error {
...
+			w.Errorf("--bump requires a local catalog: %v", catErr)
+			return exitcode.WithCode(catErr, exitcode.CatalogError)
 		}
+		bumpVer, bumpErr := solution.BumpVersion(ctx, localCat, catalog.ArtifactKindSolution, name, bumpLevel)
+		if bumpErr != nil {
+			w.Errorf("--bump: %v", bumpErr)
+			return exitcode.WithCode(bumpErr, exitcode.InvalidInput)
+		}
+		version = bumpVer
+		w.Infof("Bumped version to %s (%s)", version.String(), opts.Bump)
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/build/solution.go:343]: The `--bump` domain logic (`ParseBumpLevel`, `BumpVersion`, `ErrNoVersions`) is thoroughly unit-tested in `build_test.go` with 94.7% coverage on `BumpVersion`. Testing `--bump` through `runBuildSolution` requires a temp local catalog with a pre-existing artifact -- that is integration-test territory. The CLI wiring (flag validation, conflict checks) is covered by `TestCommandBuildSolution_BumpFlag`, `_BumpConflictsWithVersion`, `_BumpConflictsWithVersionedTag`, and `_BumpInvalidLevel`. Adding an integration test with a temp catalog is out of scope for this PR but tracked for follow-up.

File: pkg/api/endpoints/solutions.go [hunk 8]
```diff
@@ -109,16 +109,56 @@ type SolutionTestResponse struct {
...
+	// relative local paths (configs/secret, dir/../../etc/passwd). Registry
+	// hostnames always contain "." or ":" in the first segment; plain
+	// directory names do not. Block the latter to prevent path traversal.
+	if strings.Contains(path, "/") {
+		firstSegment := strings.SplitN(path, "/", 2)[0]
+		if !strings.Contains(firstSegment, ".") && !strings.Contains(firstSegment, ":") {
+			return huma.NewError(http.StatusBadRequest,
+				fmt.Sprintf("%s: path must be a URL or catalog reference, not a local file path", opName))
+		}
+	}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/api/endpoints/solutions.go:167]: Good catch — fixed. Added explicit `..` segment rejection at the top of `requireRemotePath`, before the scheme check, so paths like `ghcr.io/../../etc/passwd` are now blocked. Added test cases for both hostname-based and HTTPS-based path traversal attempts.

File: pkg/api/endpoints/solutions.go [hunk 9]
```diff
@@ -109,16 +109,56 @@ type SolutionTestResponse struct {
...
+
+	// Explicitly allowed schemes.
+	if strings.HasPrefix(lower, "http://") || strings.HasPrefix(lower, "https://") || strings.HasPrefix(lower, "oci://") {
+		return nil
+	}
+
+	// Block all other URL schemes (file://, ftp://, data://, etc.).
+	if strings.Contains(lower, "://") {
+		return huma.NewError(http.StatusBadRequest,
+			fmt.Sprintf("%s: unsupported URL scheme; only http://, https://, and oci:// are allowed", opName))
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/api/endpoints/solutions.go]: Good call — removed `oci://` from the allowed schemes. It now falls through to the "unsupported URL scheme" error. Will re-add when the loader supports OCI natively.

File: pkg/cmd/scafctl/get/solution/solution_test.go [hunk 10]
```diff
@@ -450,3 +455,249 @@ func BenchmarkCommandSolution_Structure(b *testing.B) {
...
+	output := outBuf.String()
+	assert.NotEmpty(t, output)
+}
+
+func TestListSolutions_WithLocalArtifacts(t *testing.T) {
+	// Cannot be parallel: temporarily redirects xdg.DataHome.
+	tmpDir := t.TempDir()
+	origDataHome := xdg.DataHome
+	xdg.DataHome = tmpDir
+	t.Cleanup(func() { xdg.DataHome = origDataHome })
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/get/solution/solution_test.go]: Good catch, fixed. Added a package-level `xdgMu sync.Mutex` that serialises the two tests mutating `xdg.DataHome`. The mutex is locked before mutation and unlocked in `t.Cleanup`, preventing races with parallel tests. Verified clean with `go test -race`.

File: pkg/cmd/scafctl/build/solution.go [hunk 11]
```diff
@@ -297,24 +318,43 @@ func runBuildSolution(ctx context.Context, opts *SolutionOptions) error {
...
+		localCat, catErr := catalog.NewLocalCatalog(*lgr)
+		if catErr != nil {
+			w.Errorf("--bump requires a local catalog: %v", catErr)
+			return exitcode.WithCode(catErr, exitcode.CatalogError)
 		}
+		bumpVer, bumpErr := solution.BumpVersion(ctx, localCat, catalog.ArtifactKindSolution, name, bumpLevel)
+		if bumpErr != nil {
+			w.Errorf("--bump: %v", bumpErr)
+			return exitcode.WithCode(bumpErr, exitcode.InvalidInput)
+		}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/build/solution.go:341]: Good catch, fixed. Now uses `errors.As` to check for `*solution.ErrNoVersions` — maps that to `exitcode.InvalidInput` (user needs `--version` for first publish) and all other BumpVersion errors to `exitcode.CatalogError`.
